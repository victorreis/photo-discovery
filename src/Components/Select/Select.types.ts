import { TestProps } from '../../Config/Tests/Test.types';

export type SelectOption = {
  id: string | number;
  value: string | number;
  text: string;
};

export interface RequiredSelectProps {
  /**
   * Array of options to be chosen.
   */
  options: SelectOption[];

  /**
   * Callcack function that is called when some option is selected.
   */
  onChange: (selected: string) => void;
}

export interface DefaultSelectProps {
  /**
   * Number of visible characters in the compoenent dropdown.
   * @default 25
   */
  numberOfVisibleCharacters?: number;

  /**
   * Placeholder text to be shown in the component.
   * @default 'Selec an option...'
   */
  placeholder?: string;
}

export interface OptionalSelectProps {}

export type SelectProps = RequiredSelectProps &
  DefaultSelectProps &
  OptionalSelectProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLSelectElement>, 'onChange'>;

export type SelectStyleProps = Required<DefaultSelectProps>;
