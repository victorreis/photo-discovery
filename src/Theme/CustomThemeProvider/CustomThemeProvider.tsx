import React, { useMemo, useState } from 'react';

import { ThemeProvider } from 'styled-components';

import {
  lightTheme,
  darkTheme,
  futuristicTheme,
  deepSpaceTheme,
} from '../Modes';
import { Mode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<Mode | 'default', CustomThemeType> = {
  default: deepSpaceTheme,
  light: lightTheme,
  dark: darkTheme,
  futuristic: futuristicTheme,
  deepSpace: deepSpaceTheme,
};

export const ThemeContext = React.createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<CustomThemeProviderProps> = {
  testID: 'ThemeContextProvider',
  themeName: 'deepSpace',
};

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: Mode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemeContext.Provider>
  );
};
