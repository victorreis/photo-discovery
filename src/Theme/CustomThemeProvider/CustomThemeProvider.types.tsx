import { CustomThemeType, Mode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: Mode) => void;
};

export type CustomThemeProviderProps = { testID?: string; themeName?: Mode };
