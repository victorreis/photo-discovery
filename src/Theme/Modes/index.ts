export { darkTheme } from './Dark.theme';
export { deepSpaceTheme } from './DeepSpace.theme';
export { futuristicTheme } from './Futuristic.theme';
export { lightTheme } from './Light.theme';
