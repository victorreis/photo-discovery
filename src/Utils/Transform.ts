export const hexToRgb = (hex: string) => {
  // Expand shorthand form (e.g. "03F") to full form (e.g. "0033FF")
  const shorthandRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
  const formatedHex = hex.replace(
    shorthandRegex,
    (_m, r: string, g: string, b: string) => r + r + g + g + b + b
  );

  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(formatedHex);
  return result
    ? `rgb(${parseInt(result[1] || '0', 16)}, ${parseInt(
        result[1] || '0',
        16
      )}, ${parseInt(result[1] || '0', 16)})`
    : 'rgb(0,0,0)';
};
