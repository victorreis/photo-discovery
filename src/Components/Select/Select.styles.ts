import styled from 'styled-components';

import { toPx } from '../../Utils/Transform';

export const SelectContainer = styled.select`
  cursor: pointer;
  height: 3rem;
  margin: 0.5rem;
  border-width: 1px;
  border-style: solid;
  border-color: ${({ theme }) => theme.colors.main.effect.normal}55;
  ${({ theme }) => theme.typographies.body1};
  background-color: ${({ theme }) => theme.colors.background.default.lightest};
  border-radius: ${({ theme }) => toPx(theme.borders.radius.LG)};
  box-sizing: border-box;
  padding: 0.5rem;
`;
