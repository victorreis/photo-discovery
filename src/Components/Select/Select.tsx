import { useState } from 'react';

import { TestProps } from '../../Config/Tests/Test.types';
import { SelectContainer } from './Select.styles';
import { SelectProps, DefaultSelectProps } from './Select.types';

export const selectDefaults: Required<DefaultSelectProps> &
  Required<TestProps> = {
  testID: 'Select',
  numberOfVisibleCharacters: 25,
  placeholder: 'Select an option',
};

export const Select: React.FC<SelectProps> = (props): JSX.Element => {
  const {
    testID = selectDefaults.testID,
    options,
    onChange,
    numberOfVisibleCharacters = selectDefaults.numberOfVisibleCharacters,
    placeholder = selectDefaults.placeholder,
    ...others
  } = props;

  const [selectedId, setSelectedId] = useState<string>();

  const handleSelection = (event: React.ChangeEvent<HTMLSelectElement>) => {
    onChange(event.target.value || '');
    setSelectedId(event.target.value);
  };

  const selectedOption = options.find((option) => option.id === selectedId);

  return (
    <SelectContainer
      data-testid={testID}
      onChange={handleSelection}
      title={String(selectedOption?.value)}
      value={selectedId}
      {...others}
    >
      <option defaultValue="" hidden value="">
        {placeholder}
      </option>
      <option />
      {options.map(({ text, id }) => (
        <option key={id} title={text} value={id}>
          {text.substring(0, Math.min(numberOfVisibleCharacters, text.length))}
          {text.length > numberOfVisibleCharacters && '...'}
        </option>
      ))}
    </SelectContainer>
  );
};
