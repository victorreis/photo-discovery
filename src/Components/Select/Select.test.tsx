import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { hexToRgb, toPx } from '../../Utils/Transform';
import { Select, selectDefaults } from './Select';
import { RequiredSelectProps, SelectOption, SelectProps } from './Select.types';

describe('select component tests', () => {
  const options: SelectOption[] = [
    { id: 1, value: 1, text: 'test1' },
    { id: 2, value: 2, text: 'test2' },
    { id: 3, value: 3, text: 'test3' },
  ];
  const onChange = jest.fn();

  const requiredProps: RequiredSelectProps = {
    options,
    onChange,
  };

  const setup = (props?: SelectProps) => {
    const renderRTR = () =>
      renderRTRCreator<SelectProps>(Select, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<SelectProps>(Select, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(selectDefaults.testID);

      expect(testInstance).toBeTruthy();
    });
  });

  describe('style tests', () => {
    it(`should have style the Container component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(selectDefaults.testID);

      expect(container).toHaveStyle({
        backgroundColor: hexToRgb(
          themes.default.colors.background.default.lightest
        ),
        borderRadius: toPx(themes.default.borders.radius.LG),
        ...themes.default.typographies.body1,
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
