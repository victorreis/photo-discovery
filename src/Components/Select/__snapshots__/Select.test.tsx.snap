// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select component tests snapshot tests should render correctly 1`] = `
<select
  className="SelectContainer-photo-discovery__sc-i3jpgt-0 iMvynY"
  data-testid="Select"
  onChange={[Function]}
  title="undefined"
>
  <option
    defaultValue=""
    hidden={true}
    value=""
  >
    Select an option
  </option>
  <option />
  <option
    title="test1"
    value={1}
  >
    test1
  </option>
  <option
    title="test2"
    value={2}
  >
    test2
  </option>
  <option
    title="test3"
    value={3}
  >
    test3
  </option>
</select>
`;
